Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+        SU2_MPI::Error("NdFlattener: Index out of range.", CURRENT_FUNCTION);
+      }
+      const typename Nd_type::Index new_offset = this->nd->GetIndices()[this->offset+i];
+      const typename Nd_type::Index new_size = this->nd->GetIndices()[this->offset+i+1] - new_offset;
+      return LookupType(new_size, static_cast<typename Base::Nd_Base_type*>(this->nd),new_offset);
+    }
+
+    /*! \brief Return exclusive upper bound for next index.
+     */
+    typename Nd_type::Index size() const {
+      return _size;
+    }
+  };
+
+  template<typename _Nd_type>
+  class IndexAccumulator<1,_Nd_type> : public IndexAccumulator_Base<1,_Nd_type>{
+  public:
+    static constexpr size_t N = 1;
+    using Nd_type = _Nd_type;
+    typedef IndexAccumulator_Base<1,_Nd_type> Base;
+
+    template<class ...ARGS> IndexAccumulator(ARGS... args): Base(args...) {}
+
+    /*! Return type of operator[].
+     * \details Data type of NdFlattener, but do also preserve constness.
+     */
+    using LookupType = typename Base::template conditional<
+      std::is_const<Nd_type>::value,
+      const typename Nd_type::Data,
+      typename Nd_type::Data
+    >::type;
+
+    /*! \brief Return (possibly const) reference to the corresponding data element.
+     * \param[in] i - Last index.
+     */
+    LookupType& operator[] (typename Nd_type::Index i) {
+      return this->nd->GetData() [ this->offset+i ];
+    }
+
+    /*! \brief Return (possibly const) pointer to data.
+     * \details If all indices except the last one are fixed, the corresponding data
+     * is stored contiguously. Return a pointer to the beginning of the
+     * block. If this IndexAccumulator was generated from a non-const NdFlattener, the
+     * pointer is non-const, otherwise it is const.
+     */
+    LookupType* data() {
+      return &(this->operator[](0));
+    }
+  };
+
+  template<typename _Nd_type>
+  class IndexAccumulator_Checked<1,_Nd_type> : public IndexAccumulator<1,_Nd_type>{
+  public:
+    static constexpr size_t N = 1;
+    using Nd_type = _Nd_type;
+    typedef IndexAccumulator<1,_Nd_type> Base;
+
+  protected:
+    /*! Exclusive upper bound for the next index. */
+    const typename Nd_type::Index _size;
+
+  public:
+    template<class ...ARGS>
+    IndexAccumulator_Checked(typename Nd_type::Index size, ARGS... args): Base(args...), _size(size) {}
+
+    /*! \brief Return (possibly const) reference to the corresponding data element, checking if the index is in its range.
+     * \param[in] i - Last index.
+     */
+    typename Base::LookupType& operator[] (typename Nd_type::Index i) {
+      if(i>=_size){
+        SU2_MPI::Error("NdFlattener: Index out of range.", CURRENT_FUNCTION);
+      }
+      return Base::operator[](i);
+    }
+
+    typename Nd_type::Index size() const {
+      return _size;
+    }
+  };
+
+} // namespace helpers
+
+inline helpers::NdFlattener_MPI_Environment<decltype(&(SU2_MPI::Allgatherv)), decltype(MPI_INT), decltype(SU2_MPI::GetComm())>
```

Comment:
using Something = NdFlattener_MPI_Environment<decltype(&(SU2_MPI::Allgatherv)), decltype(MPI_INT), decltype(SU2_MPI::GetComm())>
To avoid repeating the shopping list.

Comm and Type have typedefs in SU2_MPI, no need for decltype.

Reply:
I decided to remove those template arguments completely, in the context of SU2 there is no need for more generality.